src/cascadia/TerminalControl/ControlCore.cpp

```diff
@@ -0,0 +1,1349 @@
+    void ControlCore::PasteText(const winrt::hstring& hstr)
+    {
+        _terminal->WritePastedText(hstr);
+        _terminal->ClearSelection();
+        _terminal->TrySnapOnInput();
+    }
+
+    FontInfo ControlCore::GetFont() const
+    {
+        return _actualFont;
+    }
+
+    til::size ControlCore::FontSizeInDips() const
+    {
+        const til::size fontSize{ GetFont().GetSize() };
+        return fontSize.scale(til::math::rounding, 1.0f / ::base::saturated_cast<float>(_compositionScaleX));
+    }
+
+    TerminalConnection::ConnectionState ControlCore::ConnectionState() const
+    {
+        return _connection.State();
+    }
+
+    hstring ControlCore::Title()
+    {
+        hstring hstr{ _terminal->GetConsoleTitle() };
+        return hstr;
+    }
+
+    hstring ControlCore::WorkingDirectory() const
+    {
+        hstring hstr{ _terminal->GetWorkingDirectory() };
+        return hstr;
+    }
+
+    bool ControlCore::BracketedPasteEnabled() const noexcept
+    {
+        return _terminal->IsXtermBracketedPasteModeEnabled();
+    }
+
+    Windows::Foundation::IReference<winrt::Windows::UI::Color> ControlCore::TabColor() noexcept
+    {
+        auto coreColor = _terminal->GetTabColor();
+        return coreColor.has_value() ? Windows::Foundation::IReference<winrt::Windows::UI::Color>(til::color{ coreColor.value() }) :
+                                       nullptr;
+    }
+
+    til::color ControlCore::BackgroundColor() const
+    {
+        return _terminal->GetDefaultBackground();
+    }
+
+    // Method Description:
+    // - Gets the internal taskbar state value
+    // Return Value:
+    // - The taskbar state of this control
+    const size_t ControlCore::TaskbarState() const noexcept
+    {
+        return _terminal->GetTaskbarState();
+    }
+
+    // Method Description:
+    // - Gets the internal taskbar progress value
+    // Return Value:
+    // - The taskbar progress of this control
+    const size_t ControlCore::TaskbarProgress() const noexcept
+    {
+        return _terminal->GetTaskbarProgress();
+    }
+
+    int ControlCore::ScrollOffset()
+    {
+        return _terminal->GetScrollOffset();
+    }
+
+    // Function Description:
+    // - Gets the height of the terminal in lines of text. This is just the
+    //   height of the viewport.
+    // Return Value:
+    // - The height of the terminal in lines of text
+    int ControlCore::ViewHeight() const
+    {
+        const auto viewPort = _terminal->GetViewport();
+        return viewPort.Height();
+    }
+
+    // Function Description:
+    // - Gets the height of the terminal in lines of text. This includes the
+    //   history AND the viewport.
+    // Return Value:
+    // - The height of the terminal in lines of text
+    int ControlCore::BufferHeight() const
+    {
+        return _terminal->GetBufferHeight();
+    }
+
+    void ControlCore::_terminalWarningBell()
+    {
+        _WarningBellHandlers(*this, nullptr);

```

Comment:
does this happen under write lock?

Reply:
Since this can only ever be triggered by output from the connection, then the Terminal already has the write lock when calling this callback.

Reply:
K. I'm worried about making callbacks under lock. I know that the Control immediately yeets them over to the main thread, thus letting the unlock unlock . . . but what if someone forgets? Callbacks and destructors are both scary under lock. No big deal if this is how it was before ofc

Reply:
Yea, this is how this was handled before for both WarningBell and TitleChanged